client/gc.go

```diff
@@ -207,17 +116,40 @@ func (a *AllocGarbageCollector) keepUsageBelowThreshold() error {
 		a.logger.Printf("[INFO] client: garbage collecting allocation %v", alloc.ID)
 
 		// Destroy the alloc runner and wait until it exits
-		ar.Destroy()
-		select {
-		case <-ar.WaitCh():
-		case <-a.shutdownCh:
-		}
+		a.destroyAllocRunner(ar)
 	}
 	return nil
 }
 
+// destroyAllocRunner is used to destroy an allocation runner. It will acquire a
+// lock to restrict parallelism and then destroy the alloc runner, returning
+// once the allocation has been destroyed.
+func (a *AllocGarbageCollector) destroyAllocRunner(ar *AllocRunner) {
+	// Acquire the destroy lock
+	select {
+	case a.destroyCh <- struct{}{}:
+	case <-a.shutdownCh:
+		return
+	}
+
+	ar.Destroy()
+
+	select {
+	case <-ar.WaitCh():
+	case <-a.shutdownCh:
+	}
+
+	a.logger.Printf("[DEBUG] client: garbage collected %q", ar.Alloc().ID)
+
+	select {
+	case <-a.destroyCh:
+	default:
+	}
+}
+
 func (a *AllocGarbageCollector) Stop() {
 	close(a.shutdownCh)
+	close(a.destroyCh)
```

Comment:
Is this safe to call concurrently with destroyAllocRunner above? I forget whether or not a race could cause a send on a closed chan on line 130 above.